test/common/upstream/local_address_selector_integration_test.cc

```diff
@@ -0,0 +1,192 @@
+
+  // Create a config that is invalid for the ``DefaultLocalAddressSelector``.
+  config_helper_.addConfigModifier([](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    uint32_t const port_value = 1234;
+    auto bind_config = bootstrap.mutable_cluster_manager()->mutable_upstream_bind_config();
+    auto local_address_selector_config = bind_config->mutable_local_address_selector();
+    ProtobufWkt::Empty empty;
+    local_address_selector_config->mutable_typed_config()->PackFrom(empty);
+    local_address_selector_config->set_name("mock.upstream.local.address.selector");
+    bind_config->mutable_source_address()->set_address("::1");
+    bind_config->mutable_source_address()->set_port_value(port_value);
+    auto extra_source_address = bind_config->add_extra_source_addresses();
+    extra_source_address->mutable_address()->set_address("::1");
+    extra_source_address->mutable_address()->set_port_value(port_value);
+    auto extra_source_address2 = bind_config->add_extra_source_addresses();
+    extra_source_address2->mutable_address()->set_address("::2");
+    extra_source_address2->mutable_address()->set_port_value(port_value);
+  });
+
+  initialize();
+
+  codec_client_ = makeHttpConnection(makeClientConnection(lookupPort("http")));
+  Http::TestRequestHeaderMapImpl request_headers{
+      {":method", "GET"}, {":path", "/test/long/url"}, {":scheme", "http"}, {":authority", "host"}};
+
+  // Send the request headers from the client, wait until they are received upstream. When they
+  // are received, send the default response headers from upstream and wait until they are
+  // received at by client
+  auto response = sendRequestAndWaitForResponse(request_headers, 0, default_response_headers_, 0);
+
+  // Verify that we called ``getUpstreamLocalAddressImpl`` on ``TestUpstreamLocalAddressSelector``.
+  EXPECT_EQ(*num_calls, 1);
+
+  // Verify the proxied request was received upstream, as expected.
+  EXPECT_TRUE(upstream_request_->complete());
+  EXPECT_EQ(0U, upstream_request_->bodyLength());
+  // Verify the proxied response was received downstream, as expected.
+  EXPECT_TRUE(response->complete());
+  EXPECT_EQ("200", response->headers().getStatusValue());
+  EXPECT_EQ(0U, response->body().size());
+}
+
+// Verify that the bind config specified on the cluster config overrides the one
+// on the cluster manager.
+TEST_P(HttpProtocolIntegrationTest, BindConfigOverride) {
+
+  // Set up custom local address selector factory
+  std::shared_ptr<size_t> num_calls = std::make_shared<size_t>(0);
+  TestUpstreamLocalAddressSelectorFactory factory(num_calls, true);
+  Registry::InjectFactory<UpstreamLocalAddressSelectorFactory> registration(factory);
+  setUpstreamCount(2);
+
+  config_helper_.addConfigModifier([](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    // Specify a cluster with bind config.
+    bootstrap.mutable_static_resources()->mutable_clusters()->Add()->MergeFrom(
```

Comment:
If we test the override case, then we should have bind_config for the bootstrap's cluster_manager.